### PR TITLE
crocodile: Add support for generating QSPI header using mkimage

### PR DESCRIPTION
**Pull request by mamtashukla**
Add struct with Flex SPI Configuration Block and enable generating
fspi header using mkimage in device tree.
Add options in Kconfig to set FlexSPI Configuration block parameter.
Refer i.MX 8M Mini Application Processor Reference Manual for
detailed information about parameters for FlexSPI Configuration block.

### CODE REVIEW

File: include/imximage.h [hunk 1]
```diff
@@ -201,6 +201,45 @@ struct imx_header {
...
+	uint8_t sclkpadover[4];
+	uint8_t datapadover[4];
+	uint8_t dqspadover[4];
+	uint8_t timeout[4];
+	uint8_t commandInt[4];
+	uint8_t datavalid[4];
+	uint8_t busyoffset[2];
+	uint8_t busybitpolarity[2];
+	uint8_t  lut[256];
+
```
Comment (zandrey) [inline include/imximage.h:240]: No need for this empty newline here, please drop it

File: include/imximage.h [hunk 2]
```diff
@@ -201,6 +201,45 @@ struct imx_header {
...
+} __attribute__((packed)) fspi_conf;
```
Comment (zandrey) [inline include/imximage.h:241]: This is _imx-specific_ struct, should be prefixed by `imx`

File: tools/imx8mimage.c [hunk 3]
```diff
@@ -455,6 +514,14 @@ void build_image(int ofd)
 
 	header_image_off = file_off + ivt_offset;
 
+#ifdef CONFIG_FSPI_CONF_HEADER
+
+	fspi_fd = open("fspi_header.bin", O_RDWR | O_CREAT, S_IRWXU);
```
Comment (zandrey) [inline tools/imx8mimage.c:519]: NAK, hardcoded filename? Should be config parameter, so that other board users would have an option to re-define it.

**Review by zandrey — COMMENTED** [review]
I suggest we send this series upstream, so we can have more feedback on it.

It looks good, but might come to contradictions with common _U-Boot_ code. Therefore, I really suggest that we review it and send upstream first.

Besides, there are other users of this functionality that can benefit from this implementation.

Cc: @htho-lgs 
Cc: @grembeter

**Comment by mamtashukla** [comment]
v2: 
- Remove extra line  in fspi_conf struct
- Replace hard coded header file name with Kconfig option 
- Remove additional call to mkimage in  `crocodile-fspi-u-boot.dtsi`. mkimage is called to build spl.bin in `imx8mm-u-boot.dtsi` which is included in `crocodile-u-booot.dtsi` and later in `crocodile-fspu-u-boot.dtsi`. This mkimage call builds header based on check for `CONFIG_FSPI_CONF_HEADER` from Kconfig. In `crocofile-fspi-u-boot-dtsi` only file name provided via `CONFIG_FSPI_CONF_FILE`  is needed to pack with binman.

**Comment by zandrey** [comment]
I'll close this PR since changes are in the upstream in following commits:

4ae1fa338bb0 doc: board: nxp: Add instructions to boot from QSPI
eca5e000d3b9 board: freescale: Add entry for imx8mm_evk_fspi_defconfig
1c3a8e7753e7 configs: imx8mm: Define CONFIG_SYS_UBOOT_BASE for i.MX8m
99a6ff5911a6 board: freescale: Add QSPI Boot support in spl for i.MX8m
3c7ad8c40e78 configs: Add config for enabling FSPI boot option for i.MX8m
de99721c350a dts: imx8mm-uboot: Add support to pack FlexSPI Header using binman
5fe1d4b5c41f tools: mkimage: Add support to generate FlexSPI Header for i.MX8m

We'll pick those up with upcoming `v2022.11` release.

**Pull request closed without merging**